src/sv-pipeline/04_variant_resolution/scripts/clean_vcf_part5_update_records.py

```diff
@@ -0,0 +1,185 @@
+    multiallelic_out = pysam.VariantFile(mulitallelic_filename, 'w', header=normal_vcf.header)
+
+    no_variant_samples_list_file = "no_called_samples.list"
+    no_variant_samples_out = open(no_variant_samples_list_file, 'w')
+
+    for idx, record in enumerate(normal_vcf):
+        multi_del = False
+        multi_dup = False
+        gt4_copystate = False
+        gt5kb_dup = False
+        gt5kb_del = False
+        if (idx - 1) % 1000 == 0:
+            print("processed {} records".format(idx), file=sys.stderr)
+        if record.id in revised_lines_by_id:
+            record = revised_lines_by_id[record.id]
+        if record.info.get('SVTYPE', None) == 'DEL':
+            if abs(record.stop - record.pos) >= 1000:
+                sample_cn_map = {s: record.samples[s]['RD_CN'] for s in non_outlier_samples}
+                if len([s for s in sample_cn_map if (sample_cn_map[s] is not None and sample_cn_map[s] > 3)]) > vf_1:
+                    multi_del = True
+            gts = [record.samples[s]['GT'] for s in non_outlier_samples]
+            if len([gt for gt in gts if gt not in biallelic_gts]) > 0:
+                gt5kb_del = True
+            if abs(record.stop - record.pos) >= 5000:
+                if not multi_del:
+                    gt5kb_del = True
+
+        if record.info.get('SVTYPE', None) == 'DUP':
+            if abs(record.stop - record.pos) >= 1000:
+                sample_cn_map = {s: record.samples[s]['RD_CN'] for s in non_outlier_samples}
+                if len([s for s in sample_cn_map if (sample_cn_map[s] is not None and sample_cn_map[s] > 4)]) > vf_1:
+                    multi_dup = True
+                if len([x for x in Counter(sample_cn_map.values()) if (x is not None and (x < 1 or x > 4))]) > 4:
+                    gt4_copystate = True
+                if len([s for s in sample_cn_map if (sample_cn_map[s] is not None and
+                                                     (sample_cn_map[s] < 1 or sample_cn_map[s] > 4) and
+                                                     gt4_copystate)]) > vf_1:
+                    multi_dup = True
+            gts = [record.samples[s]['GT'] for s in non_outlier_samples]
+            if len([gt for gt in gts if gt not in biallelic_gts]) > 0:
+                gt5kb_dup = True
+            if abs(record.stop - record.pos) >= 5000:
+                if not multi_dup:
+                    gt5kb_dup = True
+
+        if gt5kb_del:
+            for sample in record.samples:
+                if not record.samples[sample]['GQ'] is None and record.samples[sample]['RD_CN'] >= 2:
+                    record.samples[sample]['GT'] = (0, 0)
+                elif not record.samples[sample]['GQ'] is None and record.samples[sample]['RD_CN'] == 1:
+                    record.samples[sample]['GT'] = (0, 1)
+                elif not record.samples[sample]['GQ'] is None:
+                    record.samples[sample]['GT'] = (1, 1)  # RD_CN 0 DEL
```

Comment:
```suggestion
            for sample_obj in record.samples.itervalues():
                if not sample_obj['GQ'] is None and sample_obj['RD_CN'] >= 2:
                    sample_obj['GT'] = (0, 0)
                elif not sample_obj['GQ'] is None and sample_obj['RD_CN'] == 1:
                    sample_obj['GT'] = (0, 1)
                elif not sample_obj['GQ'] is None:
                    sample_obj['GT'] = (1, 1)  # RD_CN 0 DEL
```
You can iterate over the sample record objects to avoid lots of dereferencing, which should provide a good speed-up.